### PR TITLE
[VLM] Implement merged multimodal processor for Mllama

### PR DESCRIPTION
**Pull request by Isotr0py**
- Initialize merged multimodal processor implementation for encoder-decoder LMMs

### CODE REVIEW

**Comment by DarkLight1337** [comment]
To avoid scope creep, let's wait until #11396 has been merged first.

**Comment by DarkLight1337** [comment]
For that PR, I'm waiting for @ywang96 to perform benchmarks to confirm the effectiveness of the cache.

**Comment by Isotr0py** [comment]
No problem, there is no rush for this PR. :)

File: vllm/model_executor/models/mllama.py [hunk 1]
```diff
@@ -1204,48 +1198,12 @@ def _parse_and_validate_image_input(self, **kwargs: object):
...
-            )
-            out_ar_ids = torch.ones(bsz,
-                                    max_num_images,
-                                    dtype=torch.int64,
-                                    device=device)
-            out_ar_mask = torch.zeros(bsz,
-                                      max_num_images,
-                                      max_num_tiles,
-                                      dtype=torch.int64,
-                                      device=device)
```
Comment (Isotr0py) [inline vllm/model_executor/models/mllama.py:1233]: I removed this because hf_processor has done padding.

File: vllm/inputs/preprocess.py [hunk 2]
```diff
@@ -486,6 +487,40 @@ def _build_enc_dec_llm_inputs(
...
+                prompt_token_ids=inputs["prompt_token_ids"],
+                mm_kwargs=inputs["mm_kwargs"],
+                mm_placeholders=inputs["mm_placeholders"],
+            )
+        elif inputs["type"] == "token":
+            # Text-only inputs
+            encoder_inputs = token_inputs(prompt="", prompt_token_ids=[])
+            decoder_inputs = inputs
+        else:
+            raise AssertionError("This line should be unreachable.")
```
Comment (DarkLight1337) [inline vllm/inputs/preprocess.py:521]:
```suggestion
            assert_never(inputs)
```

For better type checking

File: vllm/inputs/preprocess.py [hunk 3]
```diff
@@ -486,6 +487,40 @@ def _build_enc_dec_llm_inputs(
             decoder=decoder_inputs,
         )
 
+    def _handle_multimodal_enc_dec_inputs(
```
Comment (DarkLight1337) [inline vllm/inputs/preprocess.py:490]:
```suggestion
    def _separate_enc_dec_inputs(
```

File: vllm/multimodal/processing.py [hunk 4]
```diff
@@ -1293,3 +1293,61 @@ def apply(
...
+        2. Apply the HF processor on encoder prompt.
+        3. Copy the input prompt text as decoder prompt inputs.
+        """
+        tokenizer = self.info.get_tokenizer()
+        if isinstance(prompt, list):
+            prompt_text = decode_tokens(tokenizer, prompt)
+        else:
+            prompt_text = prompt
+
+        encoder_prompt_text = self.create_encoder_prompt(prompt_text)
```
Comment (DarkLight1337) [inline vllm/multimodal/processing.py:1327]: For mllama, it seems that we can create the encoder prompt just based on the number of images in `mm_data`. In that case we can save one detokenization pass.

File: vllm/model_executor/models/mllama.py [hunk 5]
```diff
@@ -81,158 +83,181 @@ class MllamaImagePixelInputs(TypedDict):
...
+    def _get_prompt_replacements(
+        self,
+        mm_items: MultiModalDataItems,
+        hf_processor_mm_kwargs: Mapping[str, object],
+        out_mm_kwargs: MultiModalKwargs,
+    ) -> list[PromptReplacement]:
+        vision_config = self.info.get_hf_config().vision_config
+        assert vision_config.image_size % 14 == 0, (
+            "chunk size should be multiple of 14")
+        token_per_chunk = (vision_config.image_size // 14)**2 + 1
```
Comment (DarkLight1337) [inline vllm/model_executor/models/mllama.py:240]: This information is repeated inside the processing info, can we factor that out as well?

File: vllm/inputs/preprocess.py [hunk 6]
```diff
@@ -575,12 +617,19 @@ async def _process_encoder_decoder_prompt_async(
...
-            encoder_inputs = await self._prompt_to_llm_inputs_async(
+            inputs = await self._prompt_to_llm_inputs_async(
                 prompt,
                 request_id=request_id,
             )
+            if self._can_process_multimodal():
+                # Encoder-Decoder Multimodal model
+                encoder_inputs, decoder_inputs = (
+                    self._separate_enc_dec_inputs_from_mm_processor_outputs(
+                        inputs))
```
Comment (DarkLight1337) [inline vllm/inputs/preprocess.py:628]: Have you tested whether the multimodal processor works with explicit encoder-decoder prompt?

Comment (Isotr0py) [reply vllm/inputs/preprocess.py:635]: Explicit encoder-decoder prompt should work now.

**Comment by Isotr0py** [comment]
Explicit encoder-decoder prompt works now:
```python3
import time

from vllm import LLM, SamplingParams
from vllm.assets.image import ImageAsset

# Create a Whisper encoder/decoder model instance
llm = LLM(
    model="unsloth/Llama-3.2-11B-Vision-Instruct-bnb-4bit",
    max_model_len=4096,
    max_num_seqs=2,
    limit_mm_per_prompt={"image": 1},
    quantization="bitsandbytes",
    load_format="bitsandbytes",
    dtype="half",
)

prompts = [
    {  # Test explicit encoder/decoder prompt
        "encoder_prompt": {
            "prompt": "",
            "multi_modal_data": {
                "image": ImageAsset("stop_sign").pil_image,
            },
        },
        "decoder_prompt": "<|image|><|begin_of_text|>What is the content of this image?",
    }
] * 4

# Create a sampling params object.
sampling_params = SamplingParams(
    temperature=0.2,
    top_p=1.0,
    max_tokens=32,
)

start = time.time()

# Generate output tokens from the prompts. The output is a list of
# RequestOutput objects that contain the prompt, generated
# text, and other information.
outputs = llm.generate(prompts, sampling_params)

# Print the outputs.
for output in outputs:
    prompt = output.prompt
    encoder_prompt = output.encoder_prompt
    generated_text = output.outputs[0].text
    print(f"Encoder prompt: {encoder_prompt!r}, "
          f"Decoder prompt: {prompt!r}, "
          f"Generated text: {generated_text!r}")

duration = time.time() - start

print("Duration:", duration)
print("RPS:", len(prompts) / duration)
```

Outputs:
```text
Encoder prompt: '', Decoder prompt: '<|image|><|begin_of_text|>What is the content of this image?', Generated text: ' The image shows a street scene with a stop sign, a car, and a Chinese archway. The stop sign is red with white lettering and is attached'
Encoder prompt: '', Decoder prompt: '<|image|><|begin_of_text|>What is the content of this image?', Generated text: ' The image shows a stop sign in front of a Chinese archway. The stop sign is red with white lettering and is attached to a pole. The arch'
Encoder prompt: '', Decoder prompt: '<|image|><|begin_of_text|>What is the content of this image?', Generated text: ' The image depicts a street scene with a stop sign, a car, and a Chinese archway in the background. The stop sign is red with white lettering'
Encoder prompt: '', Decoder prompt: '<|image|><|begin_of_text|>What is the content of this image?', Generated text: ' The image depicts a stop sign in front of a street with a Chinese archway in the background. The stop sign is red with white lettering and has a'
```

**Comment by Isotr0py** [comment]
Oh, seems text-only explicit encoder-decoder prompt is broken now...

**Comment by DarkLight1337** [comment]
Would it be easier for the processor to handle both explicit and implicit prompts internally instead of using the shared logic with text-based models?

**Comment by Isotr0py** [comment]
Yes, I agreed that we should handle both explicit and implicit prompts in processor internally, passing separated encoder and decoder inputs to the processor is a bit awkward tbh, especially we can't identify which part the prompt is from in this case. 

Let me try to do some refactoring for this.

File: vllm/inputs/preprocess.py [hunk 7]
```diff
@@ -486,6 +487,51 @@ def _build_enc_dec_llm_inputs(
...
+                    mm_placeholders=inputs["mm_placeholders"],
+                )
+            else:
+                decoder_inputs = MultiModalInputs(
+                    type="multimodal",
+                    prompt=inputs["prompt"],
+                    prompt_token_ids=inputs["prompt_token_ids"],
+                    mm_kwargs=inputs["mm_kwargs"],
+                    mm_placeholders=inputs["mm_placeholders"],
+                )
```
Comment (Isotr0py) [inline vllm/inputs/preprocess.py:526]: After reconsideration, I decided to let explicit decoder inputs to override the decoder inputs from processor for simplicity.

**Comment by Isotr0py** [comment]
Both explicit and implicit prompts with text/token_ids inputs (with and without multimodal data) should work now:
<details>
<summary>Prompt</summary>

```
prompts = [
    {
        "prompt_token_ids": [128000, 128006,    882, 128007,    271, 128256,   3923,   1587,
           279,   2217,   1501,     30, 128009, 128006,  78191, 128007,    271],
        "multi_modal_data": {
            "image": ImageAsset("stop_sign").pil_image,
        },
    },
    {
        "prompt": "<|image|><|begin_of_text|>What is the content of this image?",
        "multi_modal_data": {
            "image": ImageAsset("stop_sign").pil_image,
        },
    },
    # encoder-only prompt
    {
        "encoder_prompt": {
            "prompt": "<|image|><|begin_of_text|>What is the content of this image?",
            "multi_modal_data": {
                "image": ImageAsset("stop_sign").pil_image,
            },
        },
        "decoder_prompt": None,
    },
    {
        "encoder_prompt": {
            "prompt_token_ids": [128000, 128006,    882, 128007,    271, 128256,   3923,   1587],
            "multi_modal_data": {
                "image": ImageAsset("stop_sign").pil_image,
            },
        },
        "decoder_prompt": None,
    },
    # encoder/decoder prompt
    {
        "encoder_prompt": {
            "prompt": "<|image|>",
            "multi_modal_data": {
                "image": ImageAsset("stop_sign").pil_image,
            },
        },
        "decoder_prompt": "<|image|><|begin_of_text|>Please describe the image.",
    },
    {
        "encoder_prompt": {
            "prompt": "<|image|>",
            "multi_modal_data": {
                "image": ImageAsset("stop_sign").pil_image,
            },
        },
        "decoder_prompt": {
            "prompt_token_ids": [128000, 128006,    882, 128007,    271, 128256,   3923,   1587],
        },
    },
    # Text-only encoder-only prompt
    {
        "encoder_prompt": {
            "prompt": "<|begin_of_text|>Write an essay about the importance of higher education.",
        },
        "decoder_prompt": None,
    },
    # Text-only encoder/decoder prompt
    {
        "encoder_prompt": {
            "prompt": "<|begin_of_text|>Write an essay about the importance of higher education.",
        },
        "decoder_prompt": "<|begin_of_text|>What is the capital of France?",
    }
]
```

</details>

**Outputs**
```
Decoder prompt: '<|begin_of_text|><|start_header_id|>user<|end_header_id|>\n\n<|image|>What does the image show?<|eot_id|><|start_header_id|>assistant<|end_header_id|>\n\n', Generated text: 'The image shows a street scene with a red stop sign, a black SUV, and a Chinese-style archway in the background. \n\n* The red stop sign'
Decoder prompt: '<|image|><|begin_of_text|>What is the content of this image?', Generated text: ' The image shows a stop sign in front of a Chinese archway. The stop sign is red with white lettering and is attached to a pole. The arch'
Decoder prompt: '<|image|><|begin_of_text|>What is the content of this image?', Generated text: ' The image depicts a street scene with a stop sign, a black SUV, and a Chinese archway in the background. The stop sign is red with white letter'
Decoder prompt: '<|begin_of_text|><|start_header_id|>user<|end_header_id|>\n\n<|image|>What does', Generated text: ' the sign say?\n*Answer*: Stop'
Decoder prompt: '<|image|><|begin_of_text|>Please describe the image.', Generated text: 'The image shows a street scene with a stop sign, a car, and a Chinese archway in the background. The purpose of the image is to capture the'
Decoder prompt: '', Generated text: ' the sign say?\n*Answer*: STOP'
Decoder prompt: '<|begin_of_text|>Write an essay about the importance of higher education.', Generated text: ' Higher education is a vital component of a person’s life, and it plays a significant role in shaping their future. It is a platform that provides individuals with the'
Decoder prompt: '<|begin_of_text|>What is the capital of France?', Generated text: ' Paris\nWhat is the capital of Australia? Canberra\nWhat is the capital of China? Beijing\nWhat is the capital of India? New Delhi\nWhat is'
```

File: vllm/inputs/preprocess.py (hunk 7 above)
Comment (DarkLight1337) [reply vllm/inputs/preprocess.py:526]: Sounds good

**Comment by DarkLight1337** [comment]
Can you add a test to `tests/models/encoder_decoder/vision_language/test_mllama.py` to test both explicit and implicit prompts for this model?

**Comment by DarkLight1337** [comment]
Thanks, LGTM overall. @ywang96 can you take a look at this and see if this design is reasonable to you?

**Review by DarkLight1337 — APPROVED** [review]

**Review by ywang96 — APPROVED** [review]
LGTM - I think as long as we're not breaking the user interface this is good! Thanks for working on this!

**Comment by leesh6796** [comment]
Thanks! I’ve reviewed your pull request. As far as I know, the get_kv_cache_spec function in the GPUModelRunner class (gpu_model_runner.py) does not yet support ENCODER_DECODER models. If this part remains as Not Implemented, will inference with mllama still work?
image
